c/tskit/trees.c

```diff
@@ -2103,22 +2126,635 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+                }
+            }
+        }
+        u = right_child[node];
+        while (u != TSK_NULL) {
+            stack_top++;
+            stack[stack_top] = u;
+            u = left_sib[u];
+        }
+    }
+
+    mut_samples_offset = 0;
+    for (tsk_size_t s = 0; s < num_sites; s++) {
+        site = &tree_sites[s];
+
+        mut_samples_row = GET_2D_ROW(mut_samples, num_sample_chunks, mut_samples_offset);
+        mut_samples_offset += self->site_mutations_length[site->id];
+
+        out_row = GET_2D_ROW(*allele_samples, num_sample_chunks, site_offsets[site->id]);
+        get_allele_samples(site, num_sample_chunks, mut_samples_row, out_row,
+            &(*num_alleles)[site->id]);
+    }
+out:
+    tsk_safe_free(mut_samples);
+    tsk_safe_free(node_paths);
+    tsk_safe_free(stack);
+
+    return ret;
+}
+
+static int
+norm_hap_weighted(tsk_size_t state_dim, const double *hap_weights,
+    tsk_size_t TSK_UNUSED(n_a), tsk_size_t TSK_UNUSED(n_b), double *result, void *params)
+{
+    sample_count_stat_params_t args = *(sample_count_stat_params_t *) params;
+    const double *weight_row;
+    double n;
+    for (tsk_size_t k = 0; k < state_dim; k++) {
+        weight_row = GET_2D_ROW(hap_weights, 3, k);
+        n = (double) args.sample_set_sizes[k];
+        // TODO: what to do when n = 0
+        result[k] = weight_row[0] / n;
+    }
+    return 0;
+}
+
+static int
+norm_total_weighted(tsk_size_t state_dim, const double *TSK_UNUSED(hap_weights),
+    tsk_size_t n_a, tsk_size_t n_b, double *result, void *TSK_UNUSED(params))
+{
+    for (tsk_size_t k = 0; k < state_dim; k++) {
+        result[k] = 1 / (double) (n_a * n_b);
+    }
+    return 0;
+}
+
+static void
+get_all_samples_bits(tsk_bit_array_t *all_samples, tsk_size_t n, tsk_size_t n_chunks)
+{
+    const tsk_bit_array_t all = ~((tsk_bit_array_t) 0);
+    const tsk_bit_array_t remainder_samples = n % TSK_BIT_ARRAY_NUM_BITS;
+    all_samples[n_chunks - 1] = remainder_samples ? ~(all << remainder_samples) : all;
+    for (tsk_size_t i = 0; i < n_chunks - 1; i++) {
+        all_samples[i] = all;
+    }
+}
+
+static void
+sample_weights_to_bit_array(const double *weights, tsk_size_t num_sample_sets,
+    tsk_size_t num_samples, tsk_size_t num_sample_chunks, tsk_bit_array_t **sample_bits)
+{
+    const double *weight_row;
+    tsk_bit_array_t *bits_row;
+    for (tsk_bit_array_t i = 0; i < num_samples; i++) {
+        weight_row = GET_2D_ROW(weights, num_sample_sets, i);
+        for (tsk_size_t k = 0; k < num_sample_sets; k++) {
             if (weight_row[k] != 0) {
-                ret = TSK_ERR_DUPLICATE_SAMPLE;
+                bits_row = GET_2D_ROW(*sample_bits, num_sample_chunks, k);
+                tsk_add_bit_to_bit_array(bits_row, i);
+            }
+        }
+    }
+}
+
+typedef int norm_func_t(tsk_size_t state_dim, const double *hap_weights, tsk_size_t n_a,
+    tsk_size_t n_b, double *result, void *params);
+
+// TODO: should site_a/b be an id_t?
+static int
```

Comment:
This function doesn't use the entire `num_alleles` array, just those for `site_a` and `site_b`; wouldn't it be better for clarity to just pass in those? (maybe this is true for `state` also?)

Reply:
Yes, I think you're right. I can pass in pointers to the num_alleles and state array instead of the beginnings + offsets. That would definitely simplify things here.

Reply:
Well, I wasn't trying to minimize the number of variables, I was trying to make it more clear what the function uses?

Reply:
I see what you mean, so I suppose removing the offsets from within this function would simplify things. I can refactor this when I make another pass over this stuff.

Reply:
Now refactored to leave the offsets in the calling function